### ISSUE TITLE
question more than issue

### ISSUE THREAD

**Issue by keithmgould** [breeko/Simple-Reinforcement-Learning-with-Tensorflow#3]
Hey, thank you for the [rewrite](https://github.com/breeko/Simple-Reinforcement-Learning-with-Tensorflow/blob/master/Part%202%20-%20Policy-based%20Agents.ipynb)!

I'm new to reinforcement learning and was hoping to get some clarity from you. Also if you think it'd help I'd be happy to make a PR to add some comments to the code that might help other newbs like me when reading....

1. On the loss function: I don't understand why you subtract the mean and divide by std.

```
        # Determine standardized rewards
        discounted_rewards = discount_rewards(rewards, gamma)
        discounted_rewards -= discounted_rewards.mean() # why?
        discounted_rewards /= discounted_rewards.std()  # why?
```

2. On the batch: Is this just to help performance? It seems like even if we update gradients at end of every episode we still train. I tried setting batch size to 1 and things seemed to work pretty well? Any clarity here appreciated.

```
    if num_episode % batch_size == 0: # why not set batch_size to 1?
```

**Comment by breeko** [comment]
Hey, thanks for reaching out and glad you’re enjoying my write up.

1. In determining rewards I want to calculate how good an action is in terms of all the available actions. It’s not necessarily about getting the exact value of a move but a relative value of a move. By dividing by the std, you standardize the rewards so it’s easier to learn for the neural network. I’ve had issues before where the values spiral out of control (negative or positive) which leads to exploding gradients. I think doing this partially remedies that and Keeps the rewards to a similar scale. Although I have experimented with including and excluding the standardizations. By all means, feel free to experiment and see if you come up with something better.

2. I use a batch for learning gradients because it generally works better than updating on every experience individually. The issue is on line versus batch learning. There are trade offs to each, but I would search on the trade offs, experiment and use whats right and most efficient for your particular task.

A lot of these questions are design decisions that I made through experimentation. There is likely a lot of room for optimization’s but this worked for this task and is what I went with.

By all means, send over a PR. Any help is useful. Also note that I’m still actively working on section 4 so if you get there it’ll be a bit messy but I’ve been making good progress so I’ll push soon.

https://cs.stackexchange.com/questions/51260/what-are-the-advantages-of-online-learning-when-training-neural-networks
